### ISSUE TITLE
Allow researchers to easily download audio recordings

### ISSUE THREAD

**Issue by jiru** [Tatoeba/tatoeba2#2502]
**Story**

Researchers are using audio recordings from Tatoeba in their work, for example to [evaluate speech recognition models](https://github.com/facebookresearch/covost#tatoeba-evaluation-data). Currently, audio consumers need to download them manually or [craft some script](https://github.com/Tatoeba/tatoeba2/issues/2018) or rely on [outdated datasets](https://commonvoice.mozilla.org/en/datasets).

Lately we received an email from a researcher whose speech recognition model spot a few mismatches between audio and text. Even if we fix the errors, researchers won’t get the fix if they don’t get updated audio.

**Ideas**
* Periodically generate and release archives of audio (we have enough space to host them so far).
* Create and maintain a script/tool that can automatically download all the audio from Tatoeba for a given language. The tool should also allow updating after the initial download.

**Comment by LBeaudoux** [comment]
> Create and maintain a script/tool that can automatically download all the audio from Tatoeba for a given language. The tool should also allow updating after the initial download.

Maybe I could add this feature to my[ tatoebatools](https://github.com/LBeaudoux/tatoebatools) library?

**Comment by jiru** [comment]
Of course, but I am afraid that a solution based solely on the CSVs and regular HTTP download would have some limitations by design, such as:
* Slow download because the many small files (could be worked around with parallel downloads, but it might easily trigger the server rate limitations)
* How can the script know if a given audio has been updated or removed?

**Comment by LBeaudoux** [comment]
> Slow download because the many small files (could be worked around with parallel downloads, but it might easily trigger the server rate limitations)

I assumed that Tatoeba would periodically generate and release monolingual archives of audio files.

> How can the script know if a given audio has been updated or removed?

If the `created` and `modified` fields of the `audios` table are added to the `sentences_with_audio` CSV export files, then the versions of the local audio files could be compared to the versions of the files available online so that only newly created or modified files are downloaded.

**Comment by jiru** [comment]
> I assumed that Tatoeba would periodically generate and release monolingual archives of audio files.

If we are to generate such archives, then there is little need to use a dedicated download script because it’s as simple as downloading the current CSVs. Besides, as a drawbacks it becomes very hard to allow delta updates. I think we may implement either static archives or a dedicated script, not both.

> If the `created` and `modified` fields of the `audios` table are added to the `sentences_with_audio` CSV export files

Yes. There is one minor drawback however: having those field modified is no guarantee that the audio file did actually change. For example, if an admin temporarily disables audio from the sentence page for some reason, and then re-enables it, then the `created` field gets changed but the audio file doesn’t need be downloaded again.

**Comment by LBeaudoux** [comment]
Maybe Tatoeba should opt for an object storage service to manage its media files as many other websites do. 

Those services offer sync tools. For example, you can mirror files from an AWS S3 bucket to a local directory with a command like this one: 
`aws s3 sync s3://mybucket .`

**Comment by jiru** [comment]
I had a look at https://archive.org/, I think it could be a good place to host the audio that is under creative commons. They provide a [cli](https://archive.org/services/docs/api/internetarchive/cli.html) called `ia` to automate upload, and the underlying architecture is basically s3.

I considered publishing the audio under the [community audio](https://archive.org/details/opensource_audio) collection, however it looks like it’s more for audio people would normally listen to, such as music, movies, podcasts etc. The [dataset collection](https://archive.org/details/datasets) looks more appropriate. They ask not to upload more than 10k files per set, so we’d have to create zip archives. Since the license and language can only be specified on the item level (not file level), we’d rather create one item containing just one zip, for each voice/language pair.

Zipping could be tricky because CK’s dataset is about 15GB already, and we don’t have that much space to write such a big temporary zip file before upload, and `ia` does not allow "streamed" upload because it fails without a Content-Length header.

Properly setting the [language metadata](https://archive.org/services/docs/api/metadata-schema/index.html#language) would require some conversion because they use [MARC codes](https://www.loc.gov/marc/languages/), which is more or less ISO 639-2. (We use ISO 639-3 and there are subtle differences such as `fre` instead of `fra`.)

**Comment by LBeaudoux** [comment]
> we don’t have that much space to write such a big temporary zip file before upload

Moving all audio files used in production to an object-storage service would free up a lot of storage from Tatoeba server. Then, you could also zip and upload to Internet Archive from the cloud.

> they use MARC codes, which is more or less ISO 639-2

As long as you stick to the codes and don't use language names, it seems that MARC codes match ISO 639-2 bibliographic codes. In the case you chose python for your script, I recommend you [this library ](https://github.com/LBeaudoux/iso639) I created to handle language code conversion.
